### PR TITLE
[MAJOR] Authentication with Auth0 in the backend closes #85

### LINKED ISSUE #85: Create authentication and role system in backend
Epic/User story #23

### PR DESCRIPTION
**Pull request by huangs08**
Closes #85

### CODE REVIEW

File: backend/src/main/app.ts [hunk 1]
```diff
@@ -11,6 +11,11 @@ import ProjectRoute from './routes/ProjectRoute';
 import { container } from 'tsyringe';
 import { failSafeHandler, httpMiddlewareError } from './middleware/ErrorMiddleware';
 import { sequelize } from './config/sequelize';
+import { AuthenticationClient, ManagementClient } from 'auth0';
```
Comment (nasaku898) [inline backend/src/main/app.ts:14]: Linter warning, please fix.

Comment (huangs08) [reply backend/src/main/app.ts:13]: Fixed, thank you for noticing it

File: backend/src/main/middleware/JWTMiddleware.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      }
+    }
+  };
+};
+
+const getProfileRoles = async (accessToken: string): Promise<Role[]> => {
+  const authenticationClient: AuthenticationClient = container.resolve('auth0-authentication-client');
+  const managementClient: ManagementClient = container.resolve('auth0-management-client');
+
+  const user: any = await authenticationClient.getProfile(accessToken);
```
Comment (nasaku898) [inline backend/src/main/middleware/JWTMiddleware.ts:49]: We should wrap this in try catch. And reject any unexpected error.

File: backend/src/main/routes/UserRoute.ts [hunk 3]
```diff
@@ -22,6 +24,13 @@ export default class UserRoute extends CommonRoutesConfig {
   configureRoutes(): express.Application {
     this.getApp()
       .route('/users')
+      // Example of restriction on route
```
Comment (nasaku898) [inline backend/src/main/routes/UserRoute.ts:27]: We can delete this file entirely. I created it as an example but we have many files now so no need.

**Review by nasaku898 — CHANGES_REQUESTED** [review]
Overall looks good the implementation. The integration with auth0 was a good idea. It should make the application more secure!

File: backend/src/main/routes/UserRoute.ts (hunk 3 above)
Comment (Manraj45) [reply backend/src/main/routes/UserRoute.ts:27]: We should still keep it since it shows an example of how to restrict the endpoints

Comment (nasaku898) [reply backend/src/main/routes/UserRoute.ts:27]: Fair points

File: backend/src/main/middleware/JWTMiddleware.ts (hunk 2 above)
Comment (huangs08) [reply backend/src/main/middleware/JWTMiddleware.ts]: Good idea. I went ahead and did as you suggested, thanks.

**Review by nasaku898 — APPROVED** [review]

**Review by jmezaber — APPROVED** [review]
Nice!

**Pull request merged**